WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs

```diff
@@ -0,0 +1,216 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using WalletWasabi.Affiliation.Models;
+using NBitcoin;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.WabiSabi.Backend.Events;
+using WalletWasabi.Bases;
+using System.Collections.Concurrent;
+using WalletWasabi.Affiliation.Models.CoinjoinRequest;
+
+namespace WalletWasabi.Affiliation;
+
+public class CoinjoinRequestsUpdater : PeriodicRunner
+{
+	private static readonly TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(60);
+	private static readonly TimeSpan Interval = TimeSpan.FromSeconds(1);
+
+	public CoinjoinRequestsUpdater(Arena arena, ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> clients, CoinJoinRequestRequestsSigner signer)
+		  : base(Interval)
+	{
+		Arena = arena;
+		Clients = clients;
+		Signer = signer;
+
+		CoinjoinRequests = new();
+		RoundData = new();
+		RoundsToUpdate = new();
+
+		AddHandlers(arena);
+	}
+
+	private Arena Arena { get; }
+	private CoinJoinRequestRequestsSigner Signer { get; }
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<FinalizedRoundDataWithRoundId> RoundsToUpdate { get; }
```

Comment:
I may be wrong but I don't think so.

`RoundsToUpdate` is used on two places with the following call trees:
  * `CoinjoinRequestsUpdater:UpdateCoinjoinRequestsAsync` <-- `CoinjoinRequestsUpdater:ActionAsync` <-- `PeriodicRunner:ExecuteAsync` <-- `AffiliationManager:ExecuteAsync` <-- `WabiSabiCoordinator:ExecuteAsync`
  *  `CoinjoinRequestsUpdater:AddCoinjoinTransaction` <-- `CoinjoinRequestsUpdater:Arena_CoinjoinTransactionAdded` <-- `Arena:FinalizeTransaction` <-- `Arena:StepOutputRegistrationPhaseAsync` <-- `Arena:ActionAsync` <-- `PeriodicRunner:ExecuteAsync` <-- `WabiSabiCoordinator:ExecuteAsync`

So both the places are called from `WabiSabiCoordinator:ExecuteAsync` and I cannot see any place where a new thread is created.

Reply:
> So it seems to me that you have two PeriodicRunners (CoinjoinRequestsUpdater and Arena) that can touch the same data without synchronization as Queue is not synchronized.

This is true but not a problem since both the PeriodicRunners use the same thread. To be specific, they use the thread that is used [here](https://github.com/trezor/WalletWasabi/blob/onvej-sl/affiliate-server/WalletWasabi/WabiSabi/WabiSabiCoordinator.cs#L102-L110).

Reply:
Kiminuo is right here. Long running tasks, even when not created with the appropiated TaskOption, end running in a different and own thread (you can print the thead id and confirm it). In this cases there are two approaches:

* lock shared resources to gain exclusive access (used a lot in wasabi) or,
* use a worker with an inbox channel as synchronization mechanism for receiving messages from the outside.

Reply:
The test is pretty convincing. I have to rethink how `CoinjoinRequestsUpdater` works.

Reply:
> Queue is [not](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1?view=net-7.0#thread-safety) thread-safe but it looks to me that you access it from two threads really.

Fixed in https://github.com/zkSNACKs/WalletWasabi/pull/9769.